### ISSUE TITLE
vserver plugin: build fails with upcoming glibc 2.24

### ISSUE THREAD

**Issue by rubenk** [collectd/collectd#1566]
I just hit this on Rawhide:

```
vserver.c: In function 'vserver_read':
vserver.c:167:3: error: 'readdir_r' is deprecated [-Werror=deprecated-declarations]
   status = readdir_r (proc, (struct dirent *) dirent_buffer, &dent);
   ^~~~~~
In file included from ./daemon/collectd.h:196:0,
                 from vserver.c:29:
/usr/include/dirent.h:183:12: note: declared here
 extern int readdir_r (DIR *__restrict __dirp,
            ^~~~~~~~~
```

**Comment by rpv-tomsk** [comment]
https://www.sourceware.org/ml/libc-alpha/2016-02/msg00093.html

**Comment by rpv-tomsk** [comment]
The 'madwifi' plugin sysfs_iterate() function  has readdir() call, not readdir_r. The same in 'processes' plugin ( in ps_read_tasks() ).

**Comment by rubenk** [comment]
Read callbacks need to be thread-safe, see e5182de and https://collectd.org/wiki/index.php/Plugin_architecture
Only glibc deprecated readdir_r, I'm wondering what they are offering as a thread-safe replacement.

**Comment by rpv-tomsk** [comment]
There is no problems with using readdir() as thead-safe while directories are different.
Look into this page: https://www.gnu.org/software/libc/manual/html_node/Reading_002fClosing-Directory.html

>In POSIX.1-2008, readdir is not thread-safe. In the GNU C Library implementation, 
>it is safe to call readdir concurrently on different dirstreams, 
>but multiple threads accessing the same dirstream result in undefined behavior. 
>readdir_r is a fully thread-safe alternative, but suffers from poor portability (see below). 
>It is recommended that you use readdir, with external locking if multiple threads
>access the same dirstream. 

and some note about readdir_r(): 

> But in current implementations (including the GNU C Library), it is safe to call readdir 
> concurrently on different dirstreams, so there is no need to use readdir_r in most multi-threaded 
>programs. In the rare case that multiple threads need to read from the same dirstream, 
>it is still better to use readdir and external synchronization. 

Plugins 'vserver', 'madwifi' and 'processes' are not using register_complex_read(), so they needs only thread safety and does not require reentrant-safety.
As all them use different 'dirstream', no conflict occurs when using readdir().
Widely-used 'processes' plugin fully confirms what there is no problem on this.
I see no real reasons for using readdir_r() in 'vserver' plugin of current versions of Collectd.

**Comment by rubenk** [comment]
Fair enough. Want to do a PR?

**Comment by rubenk** [comment]
@rpv-tomsk , are you still interested in this?

**Comment by rpv-tomsk** [comment]
Hi! 

I'm have no time to work on this currently, sorry.
Also I don't use vserver at all.

**Comment by rubenk** [comment]
Ok, no problem!

**Comment by MostAwesomeDude** [comment]
Hi! I've opened https://github.com/NixOS/nixpkgs/pull/18956 in NixOS to patch around this. I'm cool with this patch heading upstream, but it hasn't been tested much.

**Comment by rubenk** [comment]
This was fixed by a2e870fa005e4612ed126a49fbebd8618e50da6f
